### PR TITLE
Add pipeline log output download

### LINKED ISSUE #566: [Feature] Download pipeline log
add a button to UI to download the log as file

### PR DESCRIPTION
**Pull request by lafriks**
Fixes #566

Download button is displayed only when you move mouse into log display element.

Download button will be available only when build step is finished (has end time)

attels

### CODE REVIEW

**Comment by 6543** [comment]
/home/maddl/git/own/woodpecker/web/src/components/repo/build/BuildLog.vue
  137:9  warning  Unexpected console statement  no-console

✖ 1 problem (0 errors, 1 warning)

File: web/src/assets/locales/en.json [hunk 1]
```diff
@@ -178,13 +178,15 @@
       "exit_code": "exit code {exitCode}",
       "loading": "Loading ...",
       "pipeline": "Pipeline #{buildId}",
+      "log_download_error": "There was error downloading the log file",
```
Comment (anbraten) [inline web/src/assets/locales/en.json:181]:
```suggestion
      "log_download_error": "There was an error while downloading the log file",
```

File: web/src/components/repo/build/BuildLog.vue [hunk 2]
```diff
@@ -8,7 +8,21 @@
       <Icon name="close" class="ml-auto" />
     </div>
 
-    <div class="flex flex-grow flex-col bg-gray-300 dark:bg-dark-gray-700 md:m-2 md:mt-0 md:rounded-md overflow-hidden">
+    <div
+      class="flex flex-grow flex-col bg-gray-300 dark:bg-dark-gray-700 md:m-2 md:mt-0 md:rounded-md overflow-hidden"
+      @mouseover="showActions = true"
+      @mouseleave="showActions = false"
+    >
+      <div v-show="showActions" class="block absolute top-0 right-0 z-50 mt-2 mr-4 sm:hidden md:block">
```
Comment (anbraten) [inline web/src/components/repo/build/BuildLog.vue:16]:
```suggestion
      <div v-show="showActions" class="absolute top-0 right-0 z-50 mt-2 mr-4 hidden md:flex">
```

File: web/src/components/repo/build/BuildLog.vue [hunk 3]
```diff
@@ -103,6 +122,42 @@ export default defineComponent({
     const fitAddon = ref(new FitAddon());
     const loadedLogs = ref(true);
     const autoScroll = ref(true); // TODO
+    const showActions = ref(false);
+    const downloadInProgress = ref(false);
+
+    async function download() {
+      if (!repo?.value || !build.value || !proc.value) {
+        return;
```
Comment (anbraten) [inline web/src/components/repo/build/BuildLog.vue:130]:
```suggestion
        throw new Error('The reposiotry, build or proc was undefined');
```

File: web/src/components/repo/build/BuildLog.vue [hunk 4]
```diff
@@ -103,6 +122,42 @@ export default defineComponent({
...
+      }
+      let logs;
+      try {
+        downloadInProgress.value = true;
+        logs = await apiClient.getLogs(repo.value.owner, repo.value.name, build.value.number, proc.value.pid);
+      } catch (e) {
+        console.error('Can not download log file', e);
```
Comment (anbraten) [inline web/src/components/repo/build/BuildLog.vue:137]:
```suggestion
```

**Review by anbraten — APPROVED** [review]

**Comment by lafriks** [comment]
> web/src/components/repo/build/BuildLog.vue 137:9 warning Unexpected console statement no-console
> 
> heavy_multiplication_x 1 problem (0 errors, 1 warning)

Moved that out to common notification function so that it could be reused

File: web/src/compositions/useNotifications.ts [hunk 5]
```diff
@@ -1,5 +1,20 @@
...
+  let a = args;
+  if (typeof args === 'string') {
+    a = {
+      title: args,
+      type: 'error',
+    };
+  }
+
+  notify(a);
+}
```
Comment (anbraten) [inline web/src/compositions/useNotifications.ts:18]:
```suggestion
function notifyError(err: Error, args?: NotificationsOptions | string): void {
  // eslint-disable-next-line no-console
  console.error(err);

  if (args === undefined || typeof args === 'string') {
    notify({
      title: args || err.message,
      type: 'error',
    });
    return;
  }

  notify(args);
}
```

Comment (lafriks) [reply web/src/compositions/useNotifications.ts:18]: err can not be Error as catch could also have other type

Comment (anbraten) [reply web/src/compositions/useNotifications.ts:18]: Ohh. Would the update suggestion work? 🤔

Comment (lafriks) [reply web/src/compositions/useNotifications.ts]: This looks good 👍

**Comment by 6543** [comment]
lint issue :X

File: web/src/compositions/useNotifications.ts [hunk 6]
```diff
@@ -1,5 +1,15 @@
...
+  // eslint-disable-next-line no-console
+  console.error(err);
+
+  const _args = typeof args === 'string' ? { title: args } : args;
+  const title = _args?.title || (err as Error)?.message || `${err}`;
+
+  notify({ type: 'error', ..._args, title });
+}
+
+export default () => ({ notify, notifyErr });
```
Comment (anbraten) [inline web/src/compositions/useNotifications.ts:15]:
```suggestion
export default () => ({ notify, notifyError });
```

File: web/src/compositions/useNotifications.ts [hunk 7]
```diff
@@ -1,5 +1,15 @@
...
 
-export default () => ({ notify });
+function notifyError(err: unknown, args: NotificationsOptions | string = {}): void {
+  // eslint-disable-next-line no-console
+  console.error(err);
+
+  const _args = typeof args === 'string' ? { title: args } : args;
+  const title = _args?.title || (err as Error)?.message || `${err}`;
+
+  notify({ type: 'error', ..._args, title });
```
Comment (anbraten) [inline web/src/compositions/useNotifications.ts:12]:
```suggestion
  const mArgs = typeof args === 'string' ? { title: args } : args;
  const title = mArgs?.title || (err as Error)?.message || `${err}`;

  notify({ type: 'error', ...mArgs, title });
```

File: web/src/components/repo/build/BuildLog.vue [hunk 8]
```diff
@@ -103,6 +122,41 @@ export default defineComponent({
...
+    async function download() {
+      if (!repo?.value || !build.value || !proc.value) {
+        throw new Error('The reposiotry, build or proc was undefined');
+      }
+      let logs;
+      try {
+        downloadInProgress.value = true;
+        logs = await apiClient.getLogs(repo.value.owner, repo.value.name, build.value.number, proc.value.pid);
+      } catch (e) {
+        notifications.notifyErr(e, i18n.t('repo.build.log_download_error'));
```
Comment (anbraten) [inline web/src/components/repo/build/BuildLog.vue:137]:
```suggestion
        notifications.notifyError(e, i18n.t('repo.build.log_download_error'));
```

**Comment by 6543** [comment]
> Uncaught (in promise) ReferenceError: notifyErr is not defined

**Comment by lafriks** [comment]
Rebased and fixed

**Pull request merged**